server/tabby_server/api/books.py

```diff
@@ -175,3 +176,69 @@ def _get_result_dict(books: list[google_books.Book]) -> dict:
         "results": book_dicts,
         "resultsCount": results_count,
     }
+
+
+@subapp.route("/recommendations", methods=["GET"])
+def books_recommendations() -> tuple[dict, HTTPStatus]:
+    """Gets a list of recommendations for the given set of books.
+
+    The request must contain parameters for two parallel lists.
+
+    Required args:
+        titles: List with each element being the title of each book, each
+            separated by |---|
+        authors: List with each element being the author(s) of each book, each
+            separated by |---|
+    """
+
+    # Titles and authors are separated by |---| because they might contain
+    # punctuation like commas, semicolons, etc.
+
+    # Check required args
+    titles_str = request.args.get("titles")
+    if not titles_str:
+        return {
+            "message": "Must specify 'titles' as a non-empty query parameter."
+        }, HTTPStatus.BAD_REQUEST
+    authors_str = request.args.get("authors")
+    if not authors_str:
+        return {
+            "message": "Must specify 'authors' as a non-empty query parameter."
+        }, HTTPStatus.BAD_REQUEST
+
+    # Extract each title and author
+    titles_list = [t.strip() for t in titles_str.split("|---|")]
+    authors_list = [a.strip() for a in authors_str.split("|---|")]
+    if any(t == "" for t in titles_list):
+        return {
+            "message": ("'titles' cannot have an empty element.")
+        }, HTTPStatus.BAD_REQUEST
+    if any(a == "" for a in authors_list):
+        return {
+            "message": ("'authors' cannot have an empty element.")
+        }, HTTPStatus.BAD_REQUEST
+
+    # If not equal-length lists, then return error
+    if len(titles_list) != len(authors_list):
+        return {
+            "message": (
+                "Both 'authors' and 'titles' lists must be equal in length."
+            )
+        }, HTTPStatus.BAD_REQUEST
+
+    # Get tags
+    tags_list = tags.get_tags(titles_list, authors_list)
+    if not tags_list:  # if empty, return error
+        return {
+            "message": "Unable to get tags from the given books."
+        }, HTTPStatus.BAD_REQUEST
+
+    # Get related books using list of tags
+    books = google_books.request_volumes_get(phrase=", ".join(tags_list))
```

Comment:
Is this really a good way to go about this? I'm wondering if creating an individual request for each tag might return more relevant book information. Or perhaps searching by subject instead of general phrase might yield better results?

Just seems weird to me that we're searching for a book that would somehow have all the tags we want.

But if this works, then it works I guess.

Reply:
Also, I'm pretty sure the space must go when making a request to Google Books. And since google_books.py doesn't format it, like replacing spaces with '+' or '%20' (which honestly seems like that may have been a mistake thinking about it), it should be done here pre-emptively.

Reply:
Yes, it works surprisingly well. At least Google Books doesn't return no books because of all the tags I give it. The purpose of using just one request is to make sure we don't hit the rate limit for Google Books calls.

> Also, I'm pretty sure the space must go when making a request to Google Books. And since google_books.py doesn't format it, like replacing spaces with '+' or '%20' (which honestly seems like that may have been a mistake thinking about it), it should be done here pre-emptively.

Good point. I don't think we've caught this error yet. It sort of reminds me of SQL injection. I'll try to follow up on this.

Reply:
Google Books may return results, but are they actually relevant?

But fair enough on the concern about API calls, even if we're allotted a thousand a day, I could see us run out during presentation if we're too wasteful.

On that note, maybe we should add a way to disable the google books function to avoid hitting the cap that we could call from the server. Maybe make it return pre-gen results instead and we simply Steve Jobs the presentation and act like it's dynamic all the way through.

Probably not, but its a funny thought.

Reply:
Yes, the Google Books results are pretty relevant. And yeah, it would be kind of insane if an app started generating content the moment it runs out of money.